# Market Basket Analysis
## Oleh : Febi Andika Dani Fajar Suryawan

## Studi Kasus
Diberikan data transaksi pembelian item pada suatu toko online. Data berisi rekaman waktu, nama pembeli, dan item pembelian dimana setiap pembeli maksimal membeli 3 buah item. Akan dilakukan market based analisis untuk mengetahui bagaimana asosiasi dan korelasi diantara item-item yang tersedia.

## Preparasi Data

In [2]:
# Memanggil library yang dibutuhkan
import pandas as pd
import numpy as np
from apyori import apriori

In [3]:
# Memanggil dataset
df=pd.read_csv('form.csv',delimiter=',')
df.head()

Timestamp                 Name  Item 1  Item 2     Item 3
0  2019/09/17 8:58:22 AM GMT+7  Firdaus Adi Nugroho      HP  Racket      Watch
1  2019/09/17 8:58:24 AM GMT+7               faizah      HP  Camera      Watch
2  2019/09/17 8:58:30 AM GMT+7               andrem   Watch  Camera  Music Pad
3  2019/09/17 8:58:30 AM GMT+7                laili  Camera   Watch      Mouse
4  2019/09/17 8:58:33 AM GMT+7                 Tara      HP   Watch  Music Pad

In [4]:
# Membuang kolom Timestamp, Name
data=df.drop(['Timestamp','Name'],axis=1)

In [5]:
# Membuat list dalam list dari transaksi pembelian barang
records = []
for i in range(data.shape[0]):
    records.append([str(data.values[i,j]) for j in range(data.shape[1])])
records

[['HP', 'Racket', 'Watch'],
 ['HP', 'Camera', 'Watch'],
 ['Watch', 'Camera', 'Music Pad'],
 ['Camera', 'Watch', 'Mouse'],
 ['HP', 'Watch', 'Music Pad'],
 ['Watch', 'Racket', 'Camera'],
 ['HP', 'Camera', 'Watch'],
 ['Watch', 'Camera', 'Music Pad'],
 ['Racket', 'Soap', 'Guitar'],
 ['Racket', 'Camera', 'Guitar'],
 ['Camera', 'Bag', 'nan'],
 ['Music Pad', 'Guitar', 'Camera'],
 ['Camera', 'Watch', 'nan'],
 ['Guitar', 'Camera', 'Music Pad'],
 ['Camera', 'Watch', 'Music Pad'],
 ['Camera', 'Racket', 'Guitar'],
 ['Guitar', 'Camera', 'Watch'],
 ['Guitar', 'Watch', 'nan'],
 ['Camera', 'Watch', 'nan'],
 ['Soap', 'Bag', 'Guitar'],
 ['Router', 'Bag', 'nan'],
 ['Watch', 'Racket', 'Soap'],
 ['Music Pad', 'Soap', 'Watch'],
 ['Mouse', 'Camera', 'Soap']]

Perhatikan bahwa terdapat sebuah item yang tertulis 'nan'. Hal ini berarti pada transaksi tersebut, tidak ada item yang dibeli pada pembelian ke-n dimana 'nan' tercetak.

## Market Based Analysis

Pada market based analysis kali ini, digunakan algoritma apriori untuk membuat aturan asosiasi antara tiap item. Pada program kali ini digunakan library apyori untuk membuat algoritma apriori.

In [6]:
# Menggunakan fungsi apriori untuk membuat asosiasi
association_rules = apriori(records, min_support=0.4, min_confidence=0.5,min_lift=1)
# Membuat list hasil dari algoritma apriori
association_results = list(association_rules)

In [7]:
# Menampilkan hasil asosiasi dari item 
Result=pd.DataFrame(columns=['Rule','Support','Confidence','Lift'])
for item in association_results:
    pair = item[2]
    for i in pair:
        items = str([x for x in i[0]])
        if i[3]!=1:
            Result=Result.append({'Rule':str([x for x in i[0]])+ " -> " +str([x for x in i[1]]),'Support':item[1],'Confidence':i[2],'Lift':i[3]},ignore_index=True)
Result

Rule   Support  Confidence  Lift
0  ['Camera'] -> ['Watch']  0.416667    0.625000   1.0
1  ['Watch'] -> ['Camera']  0.416667    0.666667   1.0

## Kesimpulan
Sebagai contoh, disini saya memilih nilai support terkecil adalah 0.4, nilai confidence terkecil adalah 0.5 dan nilai lift terkecil adalah 1 untuk membuat asosiasi diantara tiap item. Dapat dilihat bahwa, untuk nilai parameter tersebut hanya terdapat dua bentuk asosiasi yaitu antara Camera dan Watch. Diperoleh pula, 
- untuk rule ke-0, nilai support sebesar 0.41667 yang mana berarti peluang pembeli membeli Watch adalah sebesar 0.416667 dan nilai confidence 0.625000 yang berarti terdapat peluang sebesar 0.625000 orang akan membeli Watch pada saat dia sudah membeli .
- untuk rule ke-1, nilai support sebesar 0.41667 yang mana berarti peluang pembeli membeli Camera adalah sebesar 0.416667 dan nilai confidence 0.666667 yang berarti terdapat peluang sebesar 0.625000 orang akan membeli Watch pada saat dia sudah membeli Camera.

**Catatan:** Dapat dilakukan pergantian nilai support, confidence dan lift untuk menemukan bentuk asosiasi yang lain dari item-item tersebut. 